File: shared-libs/go/models/device.go
// Package models provides canonical data model structs for the UBR NMS.
package models

import "time"

type DeviceType string
type DeviceStatus string

const (
	DeviceTypeBTS DeviceType = "BTS"
	DeviceTypeCPE DeviceType = "CPE"
	DeviceTypeIDU DeviceType = "IDU"

	DeviceStatusOnline        DeviceStatus = "online"
	DeviceStatusOffline       DeviceStatus = "offline"
	DeviceStatusProvisioning  DeviceStatus = "provisioning"
	DeviceStatusDecommissioned DeviceStatus = "decommissioned"
)

type DeviceTag struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

type DeviceEntity struct {
	DeviceID           string       `json:"deviceId"`
	SerialNumber       string       `json:"serialNumber"`
	MacAddress         string       `json:"macAddress"`
	IPAddress          string       `json:"ipAddress,omitempty"`
	DeviceType         DeviceType   `json:"deviceType"`
	Model              string       `json:"model,omitempty"`
	FirmwareVersion    string       `json:"firmwareVersion,omitempty"`
	Region             string       `json:"region,omitempty"`
	Latitude           float64      `json:"latitude,omitempty"`
	Longitude          float64      `json:"longitude,omitempty"`
	Elevation          float64      `json:"elevation,omitempty"`
	Azimuth            int          `json:"azimuth,omitempty"`
	Tilt               int          `json:"tilt,omitempty"`
	Status             DeviceStatus `json:"status"`
	UptimeSeconds      int64        `json:"uptimeSeconds,omitempty"`
	ConnectedBTSSerial *string      `json:"connectedBtsSerial"`
	ConnectedCPECount  int          `json:"connectedCpeCount,omitempty"`
	ConnectedIDUCount  int          `json:"connectedIduCount,omitempty"`
	Tags               []DeviceTag  `json:"tags,omitempty"`
	OrganizationID     string       `json:"organizationId,omitempty"`
	NetworkID          string       `json:"networkId,omitempty"`
	CreatedAt          time.Time    `json:"createdAt,omitempty"`
	UpdatedAt          time.Time    `json:"updatedAt,omitempty"`
}

File: services/inventory-service/src/main/java/com/ubrnms/inventory/model/Device.java
package com.ubrnms.inventory.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.GeoSpatialIndexed;
import org.springframework.data.mongodb.core.index.GeoSpatialIndexType;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

/**
 * Unified device document — stores BTS (NMS-IV-02) and CPE (NMS-IV-03) fields.
 */
@Data
@NoArgsConstructor
@JsonIgnoreProperties(ignoreUnknown = true)
@Document(collection = "devices")
@CompoundIndex(name = "idx_serial", def = "{'serialNumber': 1}", unique = true)
@CompoundIndex(name = "idx_mac", def = "{'macAddress': 1}")
@CompoundIndex(name = "idx_ip", def = "{'ipAddress': 1}")
public class Device {

    @Id
    private String id;

    /** BTS or CPE */
    @Indexed
    private String deviceType;

    // --- Core fields (both BTS and CPE) ---
    @Indexed(unique = true)
    private String serialNumber;

    private String model;

    @Indexed
    private String macAddress;

    @Indexed
    private String ipAddress;

    private String firmwareVersion;
    private String softwareVersion;
    private String status;         // ACTIVE, INACTIVE, FAULTY, DECOMMISSIONED
    private long   uptimeSeconds;

    // --- Location ---
    private double latitude;
    private double longitude;
    private double elevation;
    private double azimuth;

    /** GeoJSON point for 2dsphere queries: [longitude, latitude] */
    @GeoSpatialIndexed(type = GeoSpatialIndexType.GEO_2DSPHERE)
    private double[] location; // [lon, lat]

    // --- BTS-specific (NMS-IV-02) ---
    private double  tilt;
    private String  channel;
    private String  channelBandwidth;
    private double  txPower;
    private Integer capacityPercentage;
    private List<String> connectedCpeSerials  = new ArrayList<>();
    private List<String> cascadedBtsSerials    = new ArrayList<>();

    // --- CPE-specific (NMS-IV-03) ---
    private String connectedBtsSerial;
    private Integer portOccupancy;
    private Integer capacity;
    private List<String> connectedIduSerials   = new ArrayList<>();

    // --- Organisational ---
    private String region;
    private String organizationId;

    // --- Metadata tags (NMS-IV-06) ---
    private List<DeviceTag> tags = new ArrayList<>();

    // --- Birth certificate reference ---
    private String birthCertificateId;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;
}
